### ISSUE TITLE
UIServer.detach(statsStorage) fails for trying to remove listeners twice

### ISSUE THREAD

**Issue by printomi** [deeplearning4j/deeplearning4j#6859]
# Use case
I would like to run multiple trainings, one after the other, but only display the current training on a UIServer. I think that detaching the StatsStorage from UIServer after each training is what I need, but I get an exception in this scenario. (Further plan: if this works, I would like to store the stats with my models to display them later.)
# How to reproduce the error
Insert this line:
```java
uiServer.detach(statsStorage);
```
into UIExample.java, after [this line](https://github.com/deeplearning4j/dl4j-examples/blob/460395e07d6d394c1c8228571d462fe662198944/dl4j-examples/src/main/java/org/deeplearning4j/examples/userInterface/UIExample.java#L39):
```java
net.fit(trainData);
```

# The error
This will cause the following error:
```
---java.lang.UnsupportedOperationException
	at java.util.concurrent.CopyOnWriteArrayList$COWIterator.remove(CopyOnWriteArrayList.java:1176)
	at org.deeplearning4j.ui.play.PlayUIServer.detach(PlayUIServer.java:324)
	at org.deeplearning4j.examples.userInterface.UIExample.main(UIExample.java:41)
```

# Discussion
I think that the following two lines of PlayUIServer are responsible for the above error:
https://github.com/deeplearning4j/deeplearning4j/blob/4670eee4f25cef84afca2ba2e2ebf46577968308/deeplearning4j/deeplearning4j-ui-parent/deeplearning4j-play/src/main/java/org/deeplearning4j/ui/play/PlayUIServer.java#L323
https://github.com/deeplearning4j/deeplearning4j/blob/4670eee4f25cef84afca2ba2e2ebf46577968308/deeplearning4j/deeplearning4j-ui-parent/deeplearning4j-play/src/main/java/org/deeplearning4j/ui/play/PlayUIServer.java#L324
The two lines try to remove the same element from `listeners`.
See also this line in BaseCollectionStatsStorage
https://github.com/deeplearning4j/deeplearning4j/blob/4670eee4f25cef84afca2ba2e2ebf46577968308/deeplearning4j/deeplearning4j-ui-parent/deeplearning4j-ui-model/src/main/java/org/deeplearning4j/ui/storage/BaseCollectionStatsStorage.java#L408

Should we change the code of PlayUIServer? Or is there another solution for my use case?

**Comment by printomi** [comment]
@AlexDBlack I was wrong with my first insight, and your fix did not fix all the problems of attaching and detaching. 
The two lines actually removed element from different `listeners` objects. The correct way to remove StatsStorageListener from `PlayUIServer.listeners` is 
`listeners.remove(p);` instead of `iterator.remove()`. (`iterator.remove()` was actually not supported, it threw UnsupportedOperationException, this was my original problem.)
```java
    @Override
    public synchronized void detach(StatsStorage statsStorage) {
        if (statsStorage == null)
            throw new IllegalArgumentException("StatsStorage cannot be null");
        if (!statsStorageInstances.contains(statsStorage))
            return; //No op
        boolean found = false;
        for (Iterator<Pair<StatsStorage, StatsStorageListener>> iterator = listeners.iterator(); iterator.hasNext();) {
            Pair<StatsStorage, StatsStorageListener> p = iterator.next();
            if (p.getFirst() == statsStorage) { //Same object, not equality
                statsStorage.deregisterStatsStorageListener(p.getSecond());
                listeners.remove(p);
                found = true;
            }
        }
        statsStorageInstances.remove(statsStorage);
        for (UIModule uiModule : uiModules) {
            uiModule.onDetach(statsStorage);
        }
        if (found) {
            log.info("StatsStorage instance detached from UI: {}", statsStorage);
        }
    }
```

And there is more:
I debugged my use case and found out three more problems in [TrainModule](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-ui-parent/deeplearning4j-play/src/main/java/org/deeplearning4j/ui/module/train/TrainModule.java).onDetach():
1. `currentSessionID` is not updated. I handled this with the following code modification:
```java
    @Override
    public void onDetach(StatsStorage statsStorage) {
        for (String s : knownSessionIDs.keySet()) {
            if (knownSessionIDs.get(s) == statsStorage) {
                knownSessionIDs.remove(s);
                currentSessionID = null;
                getDefaultSession();
            }
        }
    }
```
This way my code started to work: show only current/last training by detaching last before attaching new.
It was not crucial in my current code, but I think that removing fully the previous session would be nice
2. `workerIdxCount` still contains the old sessionId
3. `workerIdxToName` still contains the old sessionId
This version of code handles these too:
```java
    @Override
    public void onDetach(StatsStorage statsStorage) {
        for (String s : knownSessionIDs.keySet()) {
            if (knownSessionIDs.get(s) == statsStorage) {
                knownSessionIDs.remove(s);
                workerIdxCount.remove(s);
                workerIdxToName.remove(s);
                currentSessionID = null;
                getDefaultSession();
            }
        }
    }
```
Now, I would like to share my above described solution in a new PR. OK?

**Comment by AlexDBlack** [comment]
Good catch. Feel free to send us a pull request with that fix, it looks good to me at first glance.

**Comment by printomi** [comment]
I created a new test case based on `testUIMultipleSessions`, but it seems that I missed to include an ND4J backend. Got this exception:
```
org.nd4j.linalg.factory.Nd4jBackend$NoAvailableBackendException: Please ensure that you have an nd4j backend on your classpath. Please see: http://nd4j.org/getstarted.html
```
If I add `nd4j-native` to the pom.xml of deeplearning4j-play, I can run these tests.

But I wonder if there is a solution without adding this dependency, which is only needed by the test codes.

Shall I add this changed pom.xml also to the PR?

**Comment by printomi** [comment]
Some background info: I use IntelliJ IDEA for running the test.
I have just installed IDEA for the first time, so I would appreciate any suggestions. I opened the maven project `deeplearning4j-play` from the cloned github repo folder. 

See this screenshot of the my project setup and the selected/available maven profiles in IDEA. (I selected x86-64.)
![dl4j-intellij_idea-maven_profiles](https://user-images.githubusercontent.com/35722866/50765222-901ffd80-1275-11e9-957b-d1e703f8af0c.PNG)
